### PR TITLE
feat: Condition count buildings in area

### PR DESCRIPTION
**Pull request by UberWaffe**
fix: Inconsistent counting of buildings

### CODE REVIEW

**Comment by UberWaffe** [comment]
This PR also includes some cleanup (during which I found some bugs) of the building counting code.

File: src/building/count.c
Comment (crudelios) [inline src/building/count.c:152]: I think you should use `map_building_at` instead. This way buildings that are partially in the area may not be counted.

Do note that this would be slightly more complex as you would have to keep track of which building id's have already been added, but it's doable.

File: src/building/count.c
Comment (crudelios) [inline src/building/count.c:9]: I understand your intentions with this, but it may become annoying to maintain later on. Though I'm not sure what the best solution would be.

File: src/building/type.h
Comment (crudelios) [inline src/building/type.h:13]: I think you should set `BUILDING_ANY` to `0`, not `-1`. You can duplicate references to the same value in an enum.

**Review by crudelios — CHANGES_REQUESTED** [review]
I finally looked at this!!!

File: src/building/count.c
Comment (UberWaffe) [reply src/building/count.c:9]: I figured this was at least easier to maintain than the previous code where it basically was coded out as multiple function calls & if statements.
But any suggestions in what I should rather do I'm open to.

File: src/building/type.h
Comment (UberWaffe) [reply src/building/type.h:13]: How would I then distinguish between the two when I load the value from file?
I.e. of the saved data of the event only shows a 0, how do I know if it was meant to check for any or none?

File: src/building/count.c
Comment (UberWaffe) [reply src/building/count.c:152]: All right, I will make the change.
Probably going to take a while.
I don't have much free time at the moment.

File: src/building/type.h
Comment (crudelios) [reply src/building/type.h:13]: Would you ever want to check for none, though? You can check for any and if one is found, you know a building shouldn't be there.

File: src/building/count.c
Comment (crudelios) [reply src/building/count.c:9]: I don't have any. This isn't perfect but it will do for now!

File: src/building/type.h
Comment (UberWaffe) [reply src/building/type.h:13]: That is a good point.

File: src/building/count.c
Comment (crudelios) [reply src/building/count.c:152]: No problem, I also am very short on time right now. It shows as it took me 5 days to reply.

File: src/building/type.h
Comment (UberWaffe) [reply src/building/type.h:13]: Done.

File: src/building/count.c
Comment (UberWaffe) [reply src/building/count.c]: Changes made.

**Comment by UberWaffe** [comment]
@crudelios Note that I still need to resolve the conflicts with the main branch (fix done on counting for fort types).

**Comment by UberWaffe** [comment]
Rebase done.
Before we merge this (if you are happy with the code), let me know so I can squash and rebase.

File: src/building/count.c
Comment (crudelios) [inline src/building/count.c:154]: You don't need to alloc such a large memory array.

You know that at most you'll have as many buildings as the area to check.

So, you can set the `array_size` as:

```c 
int grid_area = abs((maxx - minx) * (maxy - miny));
int array_size = grid_area < building_count() ? grid_area : building_count();
```

Also, nitpick. Code format should be:

```c
    int *found_buildings = (int *) malloc(array_size * sizeof(int));
```

File: src/building/count.c
Comment (crudelios) [inline src/building/count.c:170]: Small performance save:

```suggestion
                for (int i = 0; i < total; i ++) {
```

File: src/building/count.c
Comment (crudelios) [inline src/building/count.c:178]: I believe `total` and `last_entry_position` will always be the same value, right? So you don't need to use both variables.

File: src/building/count.c
Comment (crudelios) [inline src/building/count.c:191]: Same thing here about the `array_size`.

File: src/building/count.c
Comment (crudelios) [inline src/building/count.c:215]: Same thing here about `total` and `last_entry_position`.

**Review by crudelios — CHANGES_REQUESTED** [review]
Just some small nitpicks and it looks good to go!

File: src/building/count.c
Comment (UberWaffe) [reply src/building/count.c:154]: Very good point. I should have thought of that.

Comment (UberWaffe) [reply src/building/count.c:154]: Fixed.

File: src/building/count.c
Comment (UberWaffe) [reply src/building/count.c:191]: Fixed.

File: src/building/count.c
Comment (UberWaffe) [reply src/building/count.c:178]: Fixed.

File: src/building/count.c
Comment (UberWaffe) [reply src/building/count.c:215]: Fixed.

File: src/building/count.c
Comment (UberWaffe) [reply src/building/count.c]: Fixed.

**Review by crudelios — APPROVED** [review]

**Pull request merged**
